Dockerfile

```diff
@@ -42,25 +43,32 @@ RUN addgroup --system --gid ${CISA_GID} ${CISA_GROUP} \
     && adduser --system --uid ${CISA_UID} --ingroup ${CISA_GROUP} ${CISA_USER}
 
 ###
-# Make sure the specified versions of pip, setuptools, and wheel are installed
+# Set up a Python virtual environment (venv); install the specified versions of pip,
+# setuptools, and wheel into it; and then install the Python dependencies for
+# the application.
 #
-# Note that we use pip3 --no-cache-dir to avoid writing to a local
+# Note that we use the --no-cache-dir flag to avoid writing to a local
 # cache.  This results in a smaller final image, at the cost of
 # slightly longer install times.
 ###
-RUN pip3 install --no-cache-dir --upgrade \
-    pip==${PYTHON_PIP_VERSION} \
-    setuptools==${PYTHON_SETUPTOOLS_VERSION} \
-    wheel==${PYTHON_WHEEL_VERSION}
+RUN python3 -m venv ${VIRTUAL_ENV} \
+    && ${VIRTUAL_ENV}/bin/python3 -m pip install --no-cache-dir --upgrade \
+        pip==${PYTHON_PIP_VERSION} \
+        setuptools==${PYTHON_SETUPTOOLS_VERSION} \
+        wheel==${PYTHON_WHEEL_VERSION} \
+    && ${VIRTUAL_ENV}/bin/python3 -m pip install --no-cache-dir --upgrade \
+        https://github.com/cisagov/skeleton-python-library/archive/v${VERSION}.tar.gz
 
 ###
-# Install Python dependencies
+# Sym-link the Python binary in the venv to the system-wide Python and add the venv to
+# the PATH.
 #
-# Note that we use pip3 --no-cache-dir to avoid writing to a local
-# cache.  This results in a smaller final image, at the cost of
-# slightly longer install times.
+# Note that we sym-link the Python binary in the venv to the system-wide Python so that
+# any calls to `python3` will use our virtual environment. We are using short flags
+# because the ln binary in Alpine Linux does not support long flags.
 ###
-RUN pip3 install --no-cache-dir https://github.com/cisagov/skeleton-python-library/archive/v${VERSION}.tar.gz
+RUN ln -sf "$(command -v python3)" "${VIRTUAL_ENV}"/bin/python3
```

Comment:
Nice trick!

Please add a short comment explaining what the `-s` and `-f` flags do here since we cannot use long options here.

Reply:
Please let me know if [77b5e34](https://github.com/cisagov/skeleton-docker/pull/190/commits/77b5e3483d2063358c20fe5b04648a9c777a8bdd) works for you.